// openclaw/dist/program-Csx0T2gv.js
import { D as tryParseLogLevel, I as getVerboseFlag, L as hasFlag, R as hasHelpOrVersion, f as isRich, j as getCommandPathWithRootOptions, p as theme, s as setVerbose, w as ALLOWED_LOG_LEVELS, z as hasRootVersionAlias } from "./globals-Bv4ZcVWM.js";
import "./paths-BfR2LXbA.js";
import { f as defaultRuntime } from "./subsystem-Cf9yS0UI.js";
import "./boolean-DTgd5CzD.js";
import { r as isTruthyEnvValue } from "./entry.js";
import "./auth-profiles-5CHn7vq1.js";
import { n as replaceCliName, r as resolveCliName } from "./command-format-BTnLVWI8.js";
import "./agent-scope-DF-nzI8H.js";
import { c as escapeRegExp } from "./utils-C5WN6czr.js";
import "./openclaw-root-DFJGXT24.js";
import "./logger-DUUyiuLB.js";
import "./exec-ByKs6PmP.js";
import "./github-copilot-token-CcBrBN3h.js";
import "./host-env-security-blJbxyQo.js";
import { t as VERSION } from "./version-Bxx5bg6l.js";
import "./registry-DoLLbW4m.js";
import "./manifest-registry-Ba187z7Z.js";
import "./dock-BgGWFqHm.js";
import "./tokens-Duelq45x.js";
import "./compact-B247y5Qt.js";
import "./plugins-BkoiCBu-.js";
import "./accounts-LPEl32gb.js";
import "./channel-config-helpers-BDId9EPL.js";
import "./send-DNQM-ubw.js";
import "./send-CY0iZqfw.js";
import "./with-timeout-CS5D2RH3.js";
import "./deliver-DjiOYDgX.js";
import "./diagnostic-Bj3zufiQ.js";
import "./accounts-_PYzdhne.js";
import "./image-ops-CQrO7d64.js";
import "./send-BE9jWQxs.js";
import "./pi-model-discovery-BU5iOFRu.js";
import "./message-channel-Be-gqLbb.js";
import "./pi-embedded-helpers-B7bn2aDM.js";
import "./sandbox-BTeDGNMJ.js";
import "./tool-catalog-CFg6jrp9.js";
import "./chrome-DmtcwLP7.js";
import "./tailscale-Bxls2k-9.js";
import "./tailnet-BuB4hnEo.js";
import "./ws-Bjkel2eG.js";
import "./auth-DgIL-BOU.js";
import "./credentials-C5dZtBqt.js";
import "./resolve-configured-secret-input-string-BGiZkpTb.js";
import "./server-context-BP-G8-mB.js";
import "./frontmatter-CDYnjLEC.js";
import "./env-overrides-BeDt30xz.js";
import "./path-alias-guards-D_v2YTP8.js";
import "./skills-B3eXdmA5.js";
import "./paths-DBHbY8ck.js";
import "./redact-SKlSb-Ph.js";
import "./errors-Duls987w.js";
import "./fs-safe-n_u5Hqlt.js";
import "./proxy-env-B1SW3C2U.js";
import "./store-kjkrRAMQ.js";
import "./ports-DOBLq-G-.js";
import "./trash-CsUKAlUm.js";
import "./server-middleware-C6_5AADb.js";
import "./sessions-DVRA-cO-.js";
import "./paths-BCX3TKIK.js";
import "./chat-envelope-B0leI413.js";
import "./tool-images-BnvE1YX0.js";
import "./thinking-DykY2Fzj.js";
import "./models-config-DXA5BkaM.js";
import "./exec-approvals-allowlist-B2KnHD_4.js";
import "./exec-safe-bin-runtime-policy-Dh7mVDQW.js";
import "./model-catalog-Ezhj-HLy.js";
import "./fetch-CvdHkSxZ.js";
import "./audio-transcription-runner-DPzqXt3G.js";
import "./fetch-guard-DpKgluWH.js";
import "./image-BCVX7urs.js";
import "./tool-display-B1jaCPlW.js";
import "./api-key-rotation-BCo2Kocu.js";
import "./proxy-fetch-4FkZfomC.js";
import "./ir-C1d-bvdn.js";
import "./render-DMmq5MTs.js";
import "./target-errors-Cq0s5cjB.js";
import "./commands-BNTk8Bcu.js";
import "./commands-registry-l51VuG0m.js";
import "./session-cost-usage-5CCZbVhO.js";
import "./session-utils-BofDsDG3.js";
import "./sqlite-Cu1wlUjK.js";
import "./call-jim2IS_j.js";
import "./pairing-token-DSWSMr10.js";
import "./fetch-C1n3F85v.js";
import "./pairing-store-CP_MX3lf.js";
import "./exec-approvals-oKbYU_E8.js";
import "./nodes-screen-ODoG4Mgx.js";
import "./restart-B-xYDDdI.js";
import "./system-run-command-j6gd3J3c.js";
import "./skill-commands-CVO7WmRA.js";
import "./pi-tools.policy-1HG57zKM.js";
import "./workspace-dirs-CquAIPLo.js";
import "./channel-activity-DMyEEcF3.js";
import "./tables-R1W5nqKL.js";
import "./server-lifecycle-BM1WkYgq.js";
import "./stagger-BeBoVYTA.js";
import "./channel-selection-CwG7Bfxc.js";
import "./plugin-auto-enable-DU_mq6in.js";
import "./send-E60uMSe6.js";
import "./outbound-attachment-wpw_WkwU.js";
import "./delivery-queue-gLCiF0so.js";
import "./send-CmHUQdYO.js";
import "./proxy-De2oME_n.js";
import "./timeouts-B6lMOqjQ.js";
import "./runtime-config-collectors-DpnTnwYq.js";
import "./command-secret-targets-BPdIQBpp.js";
import "./connection-auth-DP9lxWFW.js";
import "./onboard-helpers-BB3T2pqg.js";
import "./prompt-style-Bcr283QM.js";
import "./pairing-labels-CZVTKvUX.js";
import "./memory-cli-DkhQZF9L.js";
import "./manager-cNsoeXS5.js";
import { t as formatDocsLink } from "./links-BvlkOkWs.js";
import "./cli-utils-DNej4wZu.js";
import "./help-format-Dl2PHE-X.js";
import "./progress-fMgIMdPa.js";
import "./plugin-registry-Crk42MD2.js";
import { n as resolveCliChannelOptions } from "./channel-options-Cw1nTGJd.js";
import { t as getSubCliCommandsWithSubcommands } from "./register.subclis-B_tMG9kS.js";
import { a as registerProgramCommands, r as getCoreCliCommandsWithSubcommands } from "./command-registry-ZarkU2_Y.js";
import { r as setProgramContext } from "./program-context-BQ3tU1FN.js";
import "./ports-DD4vm6Tl.js";
import { n as formatCliBannerLine, r as hasEmittedCliBanner, t as emitCliBanner } from "./banner-Be2_SHE5.js";
import { Command, InvalidArgumentError } from "commander";
//#region src/cli/program/context.ts
function createProgramContext() {
	let cachedChannelOptions;
	const getChannelOptions = () => {
		if (cachedChannelOptions === void 0) cachedChannelOptions = resolveCliChannelOptions();
		return cachedChannelOptions;
	};
	return {
		programVersion: VERSION,
		get channelOptions() {
			return getChannelOptions();
		},
		get messageChannelOptions() {
			return getChannelOptions().join("|");
		},
		get agentChannelOptions() {
			return ["last", ...getChannelOptions()].join("|");
		}
	};
}
//#endregion
//#region src/cli/log-level-option.ts
const CLI_LOG_LEVEL_VALUES = ALLOWED_LOG_LEVELS.join("|");
function parseCliLogLevelOption(value) {
	const parsed = tryParseLogLevel(value);
	if (!parsed) throw new InvalidArgumentError(`Invalid --log-level (use ${CLI_LOG_LEVEL_VALUES})`);
	return parsed;
}
//#endregion
//#region src/cli/program/help.ts
const CLI_NAME = resolveCliName();
const CLI_NAME_PATTERN = escapeRegExp(CLI_NAME);
const ROOT_COMMANDS_WITH_SUBCOMMANDS = new Set([...getCoreCliCommandsWithSubcommands(), ...getSubCliCommandsWithSubcommands()]);
const ROOT_COMMANDS_HINT = "Hint: commands suffixed with * have subcommands. Run <command> --help for details.";
const EXAMPLES = [
	["openclaw models --help", "Show detailed help for the models command."],
	["openclaw channels login --verbose", "Link personal WhatsApp Web and show QR + connection logs."],
	["openclaw message send --target +15555550123 --message \"Hi\" --json", "Send via your web session and print JSON result."],
	["openclaw gateway --port 18789", "Run the WebSocket Gateway locally."],
	["openclaw --dev gateway", "Run a dev Gateway (isolated state/config) on ws://127.0.0.1:19001."],
	["openclaw gateway --force", "Kill anything bound to the default gateway port, then start it."],
	["openclaw gateway ...", "Gateway control via WebSocket."],
	["openclaw agent --to +15555550123 --message \"Run summary\" --deliver", "Talk directly to the agent using the Gateway; optionally send the WhatsApp reply."],
	["openclaw message send --channel telegram --target @mychat --message \"Hi\"", "Send via your Telegram bot."]
];
function configureProgramHelp(program, ctx) {
	program.name(CLI_NAME).description("").version(ctx.programVersion).option("--dev", "Dev profile: isolate state under ~/.openclaw-dev, default gateway port 19001, and shift derived ports (browser/canvas)").option("--profile <name>", "Use a named profile (isolates OPENCLAW_STATE_DIR/OPENCLAW_CONFIG_PATH under ~/.openclaw-<name>)").option("--log-level <level>", `Global log level override for file + console (${CLI_LOG_LEVEL_VALUES})`, parseCliLogLevelOption);
	program.option("--no-color", "Disable ANSI colors", false);
	program.helpOption("-h, --help", "Display help for command");
	program.helpCommand("help [command]", "Display help for command");
	program.configureHelp({
		sortSubcommands: true,
		sortOptions: true,
		optionTerm: (option) => theme.option(option.flags),
		subcommandTerm: (cmd) => {
			const hasSubcommands = cmd.parent === program && ROOT_COMMANDS_WITH_SUBCOMMANDS.has(cmd.name());
			return theme.command(hasSubcommands ? `${cmd.name()} *` : cmd.name());
		}
	});
	const formatHelpOutput = (str) => {
		let output = str;
		if (new RegExp(`^Usage:\\s+${CLI_NAME_PATTERN}\\s+\\[options\\]\\s+\\[command\\]\\s*$`, "m").test(output) && /^Commands:/m.test(output)) output = output.replace(/^Commands:/m, `Commands:\n  ${theme.muted(ROOT_COMMANDS_HINT)}`);
		return output.replace(/^Usage:/gm, theme.heading("Usage:")).replace(/^Options:/gm, theme.heading("Options:")).replace(/^Commands:/gm, theme.heading("Commands:"));
	};
	program.configureOutput({
		writeOut: (str) => {
			process.stdout.write(formatHelpOutput(str));
		},
		writeErr: (str) => {
			process.stderr.write(formatHelpOutput(str));
		},
		outputError: (str, write) => write(theme.error(str))
	});
	if (hasFlag(process.argv, "-V") || hasFlag(process.argv, "--version") || hasRootVersionAlias(process.argv)) {
		console.log(ctx.programVersion);
		process.exit(0);
	}
	program.addHelpText("beforeAll", () => {
		if (hasEmittedCliBanner()) return "";
		const rich = isRich();
		return `\n${formatCliBannerLine(ctx.programVersion, { richTty: rich })}\n`;
	});
	const fmtExamples = EXAMPLES.map(([cmd, desc]) => `  ${theme.command(replaceCliName(cmd, CLI_NAME))}\n    ${theme.muted(desc)}`).join("\n");
	program.addHelpText("afterAll", ({ command }) => {
		if (command !== program) return "";
		const docs = formatDocsLink("/cli", "docs.openclaw.ai/cli");
		return `\n${theme.heading("Examples:")}\n${fmtExamples}\n\n${theme.muted("Docs:")} ${docs}\n`;
	});
}
//#endregion
//#region src/cli/program/preaction.ts
function setProcessTitleForCommand(actionCommand) {
	let current = actionCommand;
	while (current.parent && current.parent.parent) current = current.parent;
	const name = current.name();
	const cliName = resolveCliName();
	if (!name || name === cliName) return;
	process.title = `${cliName}-${name}`;
}
const PLUGIN_REQUIRED_COMMANDS = new Set([
	"message",
	"channels",
	"directory",
	"agents",
	"configure",
	"onboard",
	"status",
	"health"
]);
const CONFIG_GUARD_BYPASS_COMMANDS = new Set([
	"doctor",
	"completion",
	"secrets"
]);
const JSON_PARSE_ONLY_COMMANDS = new Set(["config set"]);
let configGuardModulePromise;
let pluginRegistryModulePromise;
function shouldBypassConfigGuard(commandPath) {
	const [primary, secondary] = commandPath;
	if (!primary) return false;
	if (CONFIG_GUARD_BYPASS_COMMANDS.has(primary)) return true;
	if (primary === "config" && secondary === "validate") return true;
	return false;
}
function loadConfigGuardModule() {
	configGuardModulePromise ??= import("./config-guard-BKa1uuIm.js").then((n) => n.t);
	return configGuardModulePromise;
}
function loadPluginRegistryModule() {
	pluginRegistryModulePromise ??= import("./plugin-registry-Crk42MD2.js").then((n) => n.n);
	return pluginRegistryModulePromise;
}
function getRootCommand(command) {
	let current = command;
	while (current.parent) current = current.parent;
	return current;
}
function getCliLogLevel(actionCommand) {
	const root = getRootCommand(actionCommand);
	if (typeof root.getOptionValueSource !== "function") return;
	if (root.getOptionValueSource("logLevel") !== "cli") return;
	const logLevel = root.opts().logLevel;
	return typeof logLevel === "string" ? logLevel : void 0;
}
function isJsonOutputMode(commandPath, argv) {
	if (!hasFlag(argv, "--json")) return false;
	const key = `${commandPath[0] ?? ""} ${commandPath[1] ?? ""}`.trim();
	if (JSON_PARSE_ONLY_COMMANDS.has(key)) return false;
	return true;
}
function registerPreActionHooks(program, programVersion) {
	program.hook("preAction", async (_thisCommand, actionCommand) => {
		setProcessTitleForCommand(actionCommand);
		const argv = process.argv;
		if (hasHelpOrVersion(argv)) return;
		const commandPath = getCommandPathWithRootOptions(argv, 2);
		if (!(isTruthyEnvValue(process.env.OPENCLAW_HIDE_BANNER) || commandPath[0] === "update" || commandPath[0] === "completion" || commandPath[0] === "plugins" && commandPath[1] === "update")) emitCliBanner(programVersion);
		const verbose = getVerboseFlag(argv, { includeDebug: true });
		setVerbose(verbose);
		const cliLogLevel = getCliLogLevel(actionCommand);
		if (cliLogLevel) process.env.OPENCLAW_LOG_LEVEL = cliLogLevel;
		if (!verbose) process.env.NODE_NO_WARNINGS ??= "1";
		if (shouldBypassConfigGuard(commandPath)) return;
		const suppressDoctorStdout = isJsonOutputMode(commandPath, argv);
		const { ensureConfigReady } = await loadConfigGuardModule();
		await ensureConfigReady({
			runtime: defaultRuntime,
			commandPath,
			...suppressDoctorStdout ? { suppressDoctorStdout: true } : {}
		});
		if (PLUGIN_REQUIRED_COMMANDS.has(commandPath[0])) {
			const { ensurePluginRegistryLoaded } = await loadPluginRegistryModule();
			ensurePluginRegistryLoaded();
		}
	});
}
//#endregion
//#region src/cli/program/build-program.ts
function buildProgram() {
	const program = new Command();
	const ctx = createProgramContext();
	const argv = process.argv;
	setProgramContext(program, ctx);
	configureProgramHelp(program, ctx);
	registerPreActionHooks(program, ctx.programVersion);
	registerProgramCommands(program, ctx, argv);
	return program;
}
//#endregion
export { buildProgram };
