// openclaw/dist/run-main-Dd94NnjT.js
import { F as getPrimaryCommand, I as getVerboseFlag, L as hasFlag, M as getCommandPositionalsWithRootOptions, N as getFlagValue, P as getPositiveIntFlagValue, R as hasHelpOrVersion, W as isValueToken, j as getCommandPathWithRootOptions } from "./globals-Bv4ZcVWM.js";
import "./paths-BfR2LXbA.js";
import { f as defaultRuntime, r as enableConsoleCapture } from "./subsystem-Cf9yS0UI.js";
import "./boolean-DTgd5CzD.js";
import { a as normalizeEnv, c as parseCliProfileArgs, o as normalizeWindowsArgv, r as isTruthyEnvValue, s as applyCliProfileEnv } from "./entry.js";
import { Kn as loadDotEnv } from "./auth-profiles-5CHn7vq1.js";
import "./agent-scope-DF-nzI8H.js";
import "./utils-C5WN6czr.js";
import "./openclaw-root-DFJGXT24.js";
import "./logger-DUUyiuLB.js";
import "./exec-ByKs6PmP.js";
import "./github-copilot-token-CcBrBN3h.js";
import "./host-env-security-blJbxyQo.js";
import { t as VERSION } from "./version-Bxx5bg6l.js";
import "./registry-DoLLbW4m.js";
import "./manifest-registry-Ba187z7Z.js";
import "./dock-BgGWFqHm.js";
import "./tokens-Duelq45x.js";
import "./compact-B247y5Qt.js";
import "./plugins-BkoiCBu-.js";
import "./accounts-LPEl32gb.js";
import "./channel-config-helpers-BDId9EPL.js";
import "./send-DNQM-ubw.js";
import "./send-CY0iZqfw.js";
import "./with-timeout-CS5D2RH3.js";
import "./deliver-DjiOYDgX.js";
import "./diagnostic-Bj3zufiQ.js";
import "./accounts-_PYzdhne.js";
import "./image-ops-CQrO7d64.js";
import "./send-BE9jWQxs.js";
import "./pi-model-discovery-BU5iOFRu.js";
import "./message-channel-Be-gqLbb.js";
import "./pi-embedded-helpers-B7bn2aDM.js";
import "./sandbox-BTeDGNMJ.js";
import "./tool-catalog-CFg6jrp9.js";
import "./chrome-DmtcwLP7.js";
import "./tailscale-Bxls2k-9.js";
import "./tailnet-BuB4hnEo.js";
import "./ws-Bjkel2eG.js";
import "./auth-DgIL-BOU.js";
import "./credentials-C5dZtBqt.js";
import "./resolve-configured-secret-input-string-BGiZkpTb.js";
import "./server-context-BP-G8-mB.js";
import "./frontmatter-CDYnjLEC.js";
import "./env-overrides-BeDt30xz.js";
import "./path-alias-guards-D_v2YTP8.js";
import "./skills-B3eXdmA5.js";
import "./paths-DBHbY8ck.js";
import "./redact-SKlSb-Ph.js";
import { i as formatUncaughtError } from "./errors-Duls987w.js";
import "./fs-safe-n_u5Hqlt.js";
import "./proxy-env-B1SW3C2U.js";
import "./store-kjkrRAMQ.js";
import "./ports-DOBLq-G-.js";
import "./trash-CsUKAlUm.js";
import "./server-middleware-C6_5AADb.js";
import "./sessions-DVRA-cO-.js";
import "./paths-BCX3TKIK.js";
import "./chat-envelope-B0leI413.js";
import "./tool-images-BnvE1YX0.js";
import "./thinking-DykY2Fzj.js";
import "./models-config-DXA5BkaM.js";
import "./exec-approvals-allowlist-B2KnHD_4.js";
import "./exec-safe-bin-runtime-policy-Dh7mVDQW.js";
import "./model-catalog-Ezhj-HLy.js";
import "./fetch-CvdHkSxZ.js";
import { _ as installUnhandledRejectionHandler } from "./audio-transcription-runner-DPzqXt3G.js";
import "./fetch-guard-DpKgluWH.js";
import "./image-BCVX7urs.js";
import "./tool-display-B1jaCPlW.js";
import "./api-key-rotation-BCo2Kocu.js";
import "./proxy-fetch-4FkZfomC.js";
import "./ir-C1d-bvdn.js";
import "./render-DMmq5MTs.js";
import "./target-errors-Cq0s5cjB.js";
import "./commands-BNTk8Bcu.js";
import "./commands-registry-l51VuG0m.js";
import "./session-cost-usage-5CCZbVhO.js";
import "./session-utils-BofDsDG3.js";
import "./sqlite-Cu1wlUjK.js";
import "./call-jim2IS_j.js";
import "./pairing-token-DSWSMr10.js";
import "./fetch-C1n3F85v.js";
import "./pairing-store-CP_MX3lf.js";
import "./exec-approvals-oKbYU_E8.js";
import "./nodes-screen-ODoG4Mgx.js";
import "./restart-B-xYDDdI.js";
import "./system-run-command-j6gd3J3c.js";
import "./skill-commands-CVO7WmRA.js";
import "./pi-tools.policy-1HG57zKM.js";
import "./workspace-dirs-CquAIPLo.js";
import "./channel-activity-DMyEEcF3.js";
import "./tables-R1W5nqKL.js";
import "./server-lifecycle-BM1WkYgq.js";
import "./stagger-BeBoVYTA.js";
import "./channel-selection-CwG7Bfxc.js";
import "./plugin-auto-enable-DU_mq6in.js";
import "./send-E60uMSe6.js";
import "./outbound-attachment-wpw_WkwU.js";
import "./delivery-queue-gLCiF0so.js";
import "./send-CmHUQdYO.js";
import "./proxy-De2oME_n.js";
import "./timeouts-B6lMOqjQ.js";
import "./runtime-config-collectors-DpnTnwYq.js";
import "./command-secret-targets-BPdIQBpp.js";
import "./connection-auth-DP9lxWFW.js";
import "./onboard-helpers-BB3T2pqg.js";
import "./prompt-style-Bcr283QM.js";
import "./pairing-labels-CZVTKvUX.js";
import "./memory-cli-DkhQZF9L.js";
import "./manager-cNsoeXS5.js";
import "./links-BvlkOkWs.js";
import "./cli-utils-DNej4wZu.js";
import "./help-format-Dl2PHE-X.js";
import "./progress-fMgIMdPa.js";
import { t as ensureOpenClawCliOnPath } from "./path-env-CrZEPD8b.js";
import "./note-DITQCNUf.js";
import "./issue-format-DGw2-8jL.js";
import { t as ensurePluginRegistryLoaded } from "./plugin-registry-Crk42MD2.js";
import { t as assertSupportedRuntime } from "./runtime-guard-CGqgXWWP.js";
import { t as emitCliBanner } from "./banner-Be2_SHE5.js";
import "./doctor-config-flow-BjBa3YNF.js";
import { n as ensureConfigReady } from "./config-guard-BKa1uuIm.js";
import process$1 from "node:process";
import "node:url";
//#region src/cli/program/routes.ts
const routeHealth = {
	match: (path) => path[0] === "health",
	loadPlugins: (argv) => !hasFlag(argv, "--json"),
	run: async (argv) => {
		const json = hasFlag(argv, "--json");
		const verbose = getVerboseFlag(argv, { includeDebug: true });
		const timeoutMs = getPositiveIntFlagValue(argv, "--timeout");
		if (timeoutMs === null) return false;
		const { healthCommand } = await import("./health-MD58MQui.js").then((n) => n.i);
		await healthCommand({
			json,
			timeoutMs,
			verbose
		}, defaultRuntime);
		return true;
	}
};
const routeStatus = {
	match: (path) => path[0] === "status",
	loadPlugins: true,
	run: async (argv) => {
		const json = hasFlag(argv, "--json");
		const deep = hasFlag(argv, "--deep");
		const all = hasFlag(argv, "--all");
		const usage = hasFlag(argv, "--usage");
		const verbose = getVerboseFlag(argv, { includeDebug: true });
		const timeoutMs = getPositiveIntFlagValue(argv, "--timeout");
		if (timeoutMs === null) return false;
		const { statusCommand } = await import("./status-DH0rKO8q.js").then((n) => n.t);
		await statusCommand({
			json,
			deep,
			all,
			usage,
			timeoutMs,
			verbose
		}, defaultRuntime);
		return true;
	}
};
const routeSessions = {
	match: (path) => path[0] === "sessions" && !path[1],
	run: async (argv) => {
		const json = hasFlag(argv, "--json");
		const allAgents = hasFlag(argv, "--all-agents");
		const agent = getFlagValue(argv, "--agent");
		if (agent === null) return false;
		const store = getFlagValue(argv, "--store");
		if (store === null) return false;
		const active = getFlagValue(argv, "--active");
		if (active === null) return false;
		const { sessionsCommand } = await import("./sessions-Ce7q1SkW.js").then((n) => n.n);
		await sessionsCommand({
			json,
			store,
			agent,
			allAgents,
			active
		}, defaultRuntime);
		return true;
	}
};
const routeAgentsList = {
	match: (path) => path[0] === "agents" && path[1] === "list",
	run: async (argv) => {
		const json = hasFlag(argv, "--json");
		const bindings = hasFlag(argv, "--bindings");
		const { agentsListCommand } = await import("./agents-LJRe4lqY.js").then((n) => n.t);
		await agentsListCommand({
			json,
			bindings
		}, defaultRuntime);
		return true;
	}
};
const routeMemoryStatus = {
	match: (path) => path[0] === "memory" && path[1] === "status",
	run: async (argv) => {
		const agent = getFlagValue(argv, "--agent");
		if (agent === null) return false;
		const json = hasFlag(argv, "--json");
		const deep = hasFlag(argv, "--deep");
		const index = hasFlag(argv, "--index");
		const verbose = hasFlag(argv, "--verbose");
		const { runMemoryStatus } = await import("./memory-cli-DkhQZF9L.js").then((n) => n.t);
		await runMemoryStatus({
			agent,
			json,
			deep,
			index,
			verbose
		});
		return true;
	}
};
function getFlagValues(argv, name) {
	const values = [];
	const args = argv.slice(2);
	for (let i = 0; i < args.length; i += 1) {
		const arg = args[i];
		if (!arg || arg === "--") break;
		if (arg === name) {
			const next = args[i + 1];
			if (!isValueToken(next)) return null;
			values.push(next);
			i += 1;
			continue;
		}
		if (arg.startsWith(`${name}=`)) {
			const value = arg.slice(name.length + 1).trim();
			if (!value) return null;
			values.push(value);
		}
	}
	return values;
}
const routes = [
	routeHealth,
	routeStatus,
	routeSessions,
	routeAgentsList,
	routeMemoryStatus,
	{
		match: (path) => path[0] === "config" && path[1] === "get",
		run: async (argv) => {
			const positionals = getCommandPositionalsWithRootOptions(argv, {
				commandPath: ["config", "get"],
				booleanFlags: ["--json"]
			});
			if (!positionals || positionals.length !== 1) return false;
			const pathArg = positionals[0];
			if (!pathArg) return false;
			const json = hasFlag(argv, "--json");
			const { runConfigGet } = await import("./config-cli-DvV7ryHB.js");
			await runConfigGet({
				path: pathArg,
				json
			});
			return true;
		}
	},
	{
		match: (path) => path[0] === "config" && path[1] === "unset",
		run: async (argv) => {
			const positionals = getCommandPositionalsWithRootOptions(argv, { commandPath: ["config", "unset"] });
			if (!positionals || positionals.length !== 1) return false;
			const pathArg = positionals[0];
			if (!pathArg) return false;
			const { runConfigUnset } = await import("./config-cli-DvV7ryHB.js");
			await runConfigUnset({ path: pathArg });
			return true;
		}
	},
	{
		match: (path) => path[0] === "models" && path[1] === "list",
		run: async (argv) => {
			const provider = getFlagValue(argv, "--provider");
			if (provider === null) return false;
			const all = hasFlag(argv, "--all");
			const local = hasFlag(argv, "--local");
			const json = hasFlag(argv, "--json");
			const plain = hasFlag(argv, "--plain");
			const { modelsListCommand } = await import("./models-CxHdes0w.js").then((n) => n.t);
			await modelsListCommand({
				all,
				local,
				provider,
				json,
				plain
			}, defaultRuntime);
			return true;
		}
	},
	{
		match: (path) => path[0] === "models" && path[1] === "status",
		run: async (argv) => {
			const probeProvider = getFlagValue(argv, "--probe-provider");
			if (probeProvider === null) return false;
			const probeTimeout = getFlagValue(argv, "--probe-timeout");
			if (probeTimeout === null) return false;
			const probeConcurrency = getFlagValue(argv, "--probe-concurrency");
			if (probeConcurrency === null) return false;
			const probeMaxTokens = getFlagValue(argv, "--probe-max-tokens");
			if (probeMaxTokens === null) return false;
			const agent = getFlagValue(argv, "--agent");
			if (agent === null) return false;
			const probeProfileValues = getFlagValues(argv, "--probe-profile");
			if (probeProfileValues === null) return false;
			const probeProfile = probeProfileValues.length === 0 ? void 0 : probeProfileValues.length === 1 ? probeProfileValues[0] : probeProfileValues;
			const json = hasFlag(argv, "--json");
			const plain = hasFlag(argv, "--plain");
			const check = hasFlag(argv, "--check");
			const probe = hasFlag(argv, "--probe");
			const { modelsStatusCommand } = await import("./models-CxHdes0w.js").then((n) => n.t);
			await modelsStatusCommand({
				json,
				plain,
				check,
				probe,
				probeProvider,
				probeProfile,
				probeTimeout,
				probeConcurrency,
				probeMaxTokens,
				agent
			}, defaultRuntime);
			return true;
		}
	}
];
function findRoutedCommand(path) {
	for (const route of routes) if (route.match(path)) return route;
	return null;
}
//#endregion
//#region src/cli/route.ts
async function prepareRoutedCommand(params) {
	const suppressDoctorStdout = hasFlag(params.argv, "--json");
	emitCliBanner(VERSION, { argv: params.argv });
	await ensureConfigReady({
		runtime: defaultRuntime,
		commandPath: params.commandPath,
		...suppressDoctorStdout ? { suppressDoctorStdout: true } : {}
	});
	if (typeof params.loadPlugins === "function" ? params.loadPlugins(params.argv) : params.loadPlugins) ensurePluginRegistryLoaded();
}
async function tryRouteCli(argv) {
	if (isTruthyEnvValue(process.env.OPENCLAW_DISABLE_ROUTE_FIRST)) return false;
	if (hasHelpOrVersion(argv)) return false;
	const path = getCommandPathWithRootOptions(argv, 2);
	if (!path[0]) return false;
	const route = findRoutedCommand(path);
	if (!route) return false;
	await prepareRoutedCommand({
		argv,
		commandPath: path,
		loadPlugins: route.loadPlugins
	});
	return route.run(argv);
}
//#endregion
//#region src/cli/run-main.ts
function rewriteUpdateFlagArgv(argv) {
	const index = argv.indexOf("--update");
	if (index === -1) return argv;
	const next = [...argv];
	next.splice(index, 1, "update");
	return next;
}
function shouldSkipPluginCommandRegistration(params) {
	if (params.hasBuiltinPrimary) return true;
	if (!params.primary) return hasHelpOrVersion(params.argv);
	return false;
}
function shouldEnsureCliPath(argv) {
	if (hasHelpOrVersion(argv)) return false;
	const [primary, secondary] = getCommandPathWithRootOptions(argv, 2);
	if (!primary) return true;
	if (primary === "status" || primary === "health" || primary === "sessions") return false;
	if (primary === "config" && (secondary === "get" || secondary === "unset")) return false;
	if (primary === "models" && (secondary === "list" || secondary === "status")) return false;
	return true;
}
async function runCli(argv = process$1.argv) {
	let normalizedArgv = normalizeWindowsArgv(argv);
	const parsedProfile = parseCliProfileArgs(normalizedArgv);
	if (!parsedProfile.ok) throw new Error(parsedProfile.error);
	if (parsedProfile.profile) applyCliProfileEnv({ profile: parsedProfile.profile });
	normalizedArgv = parsedProfile.argv;
	loadDotEnv({ quiet: true });
	normalizeEnv();
	if (shouldEnsureCliPath(normalizedArgv)) ensureOpenClawCliOnPath();
	assertSupportedRuntime();
	if (await tryRouteCli(normalizedArgv)) return;
	enableConsoleCapture();
	const { buildProgram } = await import("./program-Csx0T2gv.js");
	const program = buildProgram();
	installUnhandledRejectionHandler();
	process$1.on("uncaughtException", (error) => {
		console.error("[openclaw] Uncaught exception:", formatUncaughtError(error));
		process$1.exit(1);
	});
	const parseArgv = rewriteUpdateFlagArgv(normalizedArgv);
	const primary = getPrimaryCommand(parseArgv);
	if (primary) {
		const { getProgramContext } = await import("./program-context-BQ3tU1FN.js").then((n) => n.n);
		const ctx = getProgramContext(program);
		if (ctx) {
			const { registerCoreCliByName } = await import("./command-registry-ZarkU2_Y.js").then((n) => n.t);
			await registerCoreCliByName(program, ctx, primary, parseArgv);
		}
		const { registerSubCliByName } = await import("./register.subclis-B_tMG9kS.js").then((n) => n.a);
		await registerSubCliByName(program, primary);
	}
	if (!shouldSkipPluginCommandRegistration({
		argv: parseArgv,
		primary,
		hasBuiltinPrimary: primary !== null && program.commands.some((command) => command.name() === primary)
	})) {
		const { registerPluginCliCommands } = await import("./cli-B0zvOW_M.js");
		const { loadValidatedConfigForPluginRegistration } = await import("./register.subclis-B_tMG9kS.js").then((n) => n.a);
		const config = await loadValidatedConfigForPluginRegistration();
		if (config) registerPluginCliCommands(program, config);
	}
	await program.parseAsync(parseArgv);
}
//#endregion
export { runCli };
